core/src/toga/sources/tree_source.py

```diff
@@ -41,27 +47,28 @@ def __repr__(self):
     # Methods required by the TreeSource interface
     ######################################################################
 
-    def __getitem__(self, index: int) -> Node:
+    def __getitem__(self, index: int) -> Node[T]:
         if self._children is None:
             raise ValueError(f"{self} is a leaf node")
 
         return self._children[index]
 
-    def __delitem__(self, index: int):
+    def __delitem__(self, index: int) -> None:
         if self._children is None:
             raise ValueError(f"{self} is a leaf node")
 
         child = self._children[index]
         del self._children[index]
 
-        # Child isn't part of this source, or a child of this node any more.
+        # Child isn't part of this source, or a child of this node anymore.
         child._parent = None
-        child._source = None
+        # TODO:PR: consider del?
+        child._source = None  # type: ignore[assignment]
```

Comment:
more questions about Optionalized attributes 🙃 

Might it be more appropriate to just delete the removed child's `Source` instead of setting it to `None`?

Setting it `None` gets rid of its reference and all....but it definitely confuses typing...

Reply:
I guess that could also work - but it means you need to guard any reference to `_source` with `hasattr` or `except AttributeError`, rather than just `is None` checks.

That said, I'm not sure we're doing much by way of `is None` checks either, so maybe it won't matter?

Reply:
> That said, I'm not sure we're doing much by way of `is None` checks either, so maybe it won't matter?

Exactly; if the `None` checks already existed, then I could just type `_source` as `TreeSource | None` and move on....but currently, the code is implicitly implemented such that `_source` will never be none in any "interesting" states of a `Node`....or so it's believed ;)

Reply:
thinking just a bit more about this and what you said....deleting the attribute with `del` doesn't actually change the status quo at all...the code is still vulnerable to dereferencing `_source` when it is "invalid".

So, to that end, is there an explicit reason to set `_source` to `None`?

Is this just to get rid of the reference? and if so, could a `Node` reasonably stick around in memory beyond its `Source` thereby keeping the `Source` around intentionally...?

This also feels like a lot of "deep" thinking for something that isn't a problem...

Reply:
The reason for a node having a source is for signalling purposes - if the node is changed, the source needs to be notified so that it can inform any listeners. If the node has been removed... I can't think of a reason it should be signalling anyone. 

There's also the case of initial construction. Node is assigned the source attribute as a result of being added to a tree; but at time of initial construction, it doesn't have a source. It's entirely plausible you might create a bunch of nodes for some reason, but not actually add them to a tree until they're needed of some reason; in that case having an explicit source (or clear expectation that _source is None) is probably desirable.